EpiAware/src/observation-processes.jl

```diff
@@ -1,54 +1,57 @@
+abstract type AbstractObservationModel end
+
+struct DelayObservations{T <: AbstractFloat, S <: Sampleable} <: AbstractObservationModel
+    delay_kernel::SparseMatrixCSC{T, Integer}
+    neg_bin_cluster_factor_prior::S
+
+    function DelayObservations(delay_int,
+            time_horizon,
+            neg_bin_cluster_factor_prior)
+        @assert all(delay_int .>= 0) "Delay interval must be non-negative"
+        @assert sum(delay_int)≈1 "Delay interval must sum to 1"
+
+        K = generate_observation_kernel(delay_int, time_horizon)
+
+        new{eltype(K), typeof(neg_bin_cluster_factor_prior)}(K,
+            neg_bin_cluster_factor_prior)
+    end
+
+    function DelayObservations(;
+            delay_distribution::ContinuousDistribution,
+            time_horizon::Integer,
+            neg_bin_cluster_factor_prior::Sampleable,
+            D_delay,
+            Δd = 1.0)
+        delay_int = create_discrete_pmf(delay_distribution; Δd = Δd, D = D_delay)
+        return DelayObservations(delay_int, time_horizon, neg_bin_cluster_factor_prior)
+    end
+end
+
 function default_delay_obs_priors()
     return (:neg_bin_cluster_factor_prior => Gamma(3, 0.05 / 3),) |> Dict
 end
 
-@model function delay_observations(
+function generate_observations(observation_model::AbstractObservationModel,
```

Comment:
From an organisational stand point does it make more sense to define a abstract type and then any types that inherit from it or define the abstract type and methods that are dispatched from it and then the specific instances of those types and methods?